### ISSUE TITLE
Investigation: assembling documents seems slow

### ISSUE THREAD

**Issue by nonprofittechy** [SuffolkLITLab/docassemble-AssemblyLine#36]
Using the ALDocument class seems to slow down document assembly. Maybe something to do with it raising more exceptions than normal attachment assembly would?

Can we find a way to test if this is correct and speed it up if it really is a problem?

**Comment by plocket** [comment]
I've found document display through ALDocument to be much slower than a built-in `attachment`. For a while, the table doubled the time because it was building each bundle twice - once for download and once for view - but I stopped that duplication of work. It still remained significantly slower than the built-in `attachment`.

**Comment by nonprofittechy** [comment]
So that's two anecdotes anyway. It would be nice to isolate and fix it.

**Comment by purplesky2016** [comment]
If we move some of the files to the Weaver's download package instead of generating them in runtime in AL, it would reduce ALDocument's workload thus speed it up. This approach could also solve a bunch of file related open issues.

**Comment by nonprofittechy** [comment]
Jonathan just added a bunch of changes in DA 1.2.35 that we should experiment with. They might do a safer job of handling exceptions w/o slowing down document generation

**Comment by nonprofittechy** [comment]
We might consider also generating the document in the background. We will need to trigger document generation in the interview order block that the weaver generates.

There may be other places where we can trigger definitions in interview order block to help reduce code running multiple times. That would happen in the weaver too.

**Comment by plocket** [comment]
[Changelog showing 1.2.35](https://docassemble.org/docs/changelog.html)

**Comment by nonprofittechy** [comment]
I think a major factor here is idempotency. If we trigger all of the "metadata" type variables of the addendum early, it could help. E.g.: trigger .enabled, .has_addendum, .overflow_fields, etc. so they are defined before the list starts getting generated. A lot of those variables won't be defined until needed.

**Comment by nonprofittechy** [comment]
See #84 for probably the best overall resolution of this.

**Comment by BryceStevenWilley** [comment]
I decided to dive into actual profiling results for this issue. It just seems really strange that it should take 10 seconds to assemble some relatively simple PDFs. After some setup, I now have actual profiling results on this. This mostly would drive improvements upstream in DA core, but maybe we can deduce some other workarounds
 out of this anyway.

## Results

The below graph is the call chain of functions that happen when `as_pdf()` is called. Each box is a function, the `%` is the percentage of time spent calling that function over the whole time profiled, and the `__ x` is the number of times that function is called. 

![pyprof](https://user-images.githubusercontent.com/6252212/122115029-f7045200-cdf1-11eb-8868-612e29baca1a.png)

Here's the top functions where the most time is spent.

![Screenshot from 2021-06-15 15-58-15](https://user-images.githubusercontent.com/6252212/122115650-b527db80-cdf2-11eb-90d3-f1910847aa84.png)

It's definitely weird that `inspect.getmodule` is called so much, and that it spends so much time there. I'll have to take a look at that next when I have time.

## Profile setup**

In  any interview, add the following sections:
```
modules:
  - cProfile
  - pstats
---
objects:
  - download_file: DAFile
---
code: |
  download_file.initialize(filename='test.prof')
  # Put whatever python expression you want to profile here in the quotes
  runctx('al_user_bundle.as_pdf()', globals=None, locals=locals(), filename=download_file.path())
---
# Somehow later on, allow the user to download the file
  download_file.url_for()
```

Then run the interview, and download the profile file. Locally, have both [kcachegrind](https://kcachegrind.github.io/html/Home.html) and [pyprof2calltree](https://pypi.org/project/pyprof2calltree/) installed.
```bash
sudo apt install kcachegrind
pip install pyprof2calltree

pyprof2calltree -i ~/Downloads/test.prof -k
```


** I accidentally overwrote all of the setup code I had, so this might not work as written. But it's close.

**Comment by nonprofittechy** [comment]
This is very interesting, Bryce. I am not sure what inspect.getmodule does, but the number of times it is called definitely screams some idempotency issue to me. I think we can probably reduce the number of exceptions that are triggered by doing things higher up the stack.

**Comment by nonprofittechy** [comment]
```
qsteenhuis@GBLS-0783457:~/docassemble$ grep -r "inspect.get" -n
docassemble_base/docassemble/base/mako/ext/turbogears.py:34:        for kw in compat.inspect_getargspec(Template.__init__)[0]:
docassemble_base/docassemble/base/mako/doc/_sources/changelog.txt:40:      additionally replace the use of ``inspect.getargspec()`` under
docassemble_base/docassemble/base/mako/doc/changelog.html:235:additionally replace the use of <code class="docutils literal"><span class="pre">inspect.getargspec()</span></code> under
docassemble_base/docassemble/base/mako/doc/build/changelog.rst:40:      additionally replace the use of ``inspect.getargspec()`` under
docassemble_base/docassemble/base/mako/compat.py:23:    from inspect import getfullargspec as inspect_getfullargspec
docassemble_base/docassemble/base/mako/compat.py:25:    def inspect_getargspec(func):
docassemble_base/docassemble/base/mako/compat.py:27:            *inspect_getfullargspec(func)[0:4]
docassemble_base/docassemble/base/mako/compat.py:30:    from inspect import getargspec as inspect_getargspec  # noqa
docassemble_base/docassemble/base/mako/compat.py:174:        return inspect.getargspec(fn)
docassemble_webapp/docassemble/webapp/server.py:3002:            source_file = inspect.getsourcefile(the_name)
docassemble_webapp/docassemble/webapp/server.py:3111:                    pg_code_cache[val] = {'doc': noquotetrunc(inspect.getdoc(user_dict[val])), 'name': str(val), 'insert': str(val) + '()', 'tag': str(val) + str(inspect.formatargspec(*inspect.getargspec(user_dict[val]))), 'git': source_code_url(user_dict[val])}
docassemble_webapp/docassemble/webapp/server.py:3120:                    pg_code_cache[val] = {'doc': noquotetrunc(inspect.getdoc(user_dict[val])), 'name': str(val), 'insert': str(val), 'git': source_code_url(user_dict[val], datatype='module')}
docassemble_webapp/docassemble/webapp/server.py:3133:                    methods = inspect.getmembers(user_dict[val], predicate=lambda x: public_method(x, user_dict[val]))
docassemble_webapp/docassemble/webapp/server.py:3139:                        method_list.append({'insert': '.' + str(name) + '()', 'name': str(name), 'doc': noquotetrunc(inspect.getdoc(value)), 'tag': '.' + str(name) + str(inspect.formatargspec(*inspect.getargspec(value))), 'git': source_code_url(value)})
docassemble_webapp/docassemble/webapp/server.py:3143:                    pg_code_cache[val] = {'doc': noquotetrunc(inspect.getdoc(user_dict[val])), 'name': str(val), 'insert': str(val), 'bases': bases, 'methods': method_list, 'git': source_code_url(user_dict[val], datatype='class')}
```

And on line 3023 of server.py, it looks like most of the calls to inspect.* are in get_vars_in_use. I think that would definitely be used for assembling a DOCX template. I wonder if it's also used when `skip undefined: True` is turned on for a PDF.

**Comment by BryceStevenWilley** [comment]
I'm not sure if idempotency is it: the `defined_fields` method is only called 4 times and `defined` and `value` are called a combined 48 times (at least in the one `as_pdf` call I profiled), which doesn't seem bad? I did it on the AL Test interview, so I think there about 20 fields in there now, so the number of fields roughly matches the times `value_if_defined` is called on our side. I think [`inspect.stack`](https://github.com/jhpyle/docassemble/blob/4ccd5afdab7efe21e156f7663791d1f6f8cc0214/docassemble_base/docassemble/base/functions.py#L3976) is incredibly expensive though, which is mentioned briefly in the [DA documentation](https://docassemble.org/docs/functions.html#value):

> The value() function is relatively inefficient. If you can use regular Python expressions instead of value(), you should do so.

There looks to be quite a few speed ups in core, that I need to test out a bit before trying to push them upstream:
* Instead of `stack()` grabbing the entire stack at once, [just call `current_frame()`](https://stackoverflow.com/a/42636264), and go up one at a time
* Using `stack(context=0)` [might be much faster since it doesn't need to read in file info](https://stackoverflow.com/a/17407257)
* [`defined()`](https://github.com/jhpyle/docassemble/blob/4ccd5afdab7efe21e156f7663791d1f6f8cc0214/docassemble_base/docassemble/base/functions.py#L3964) and [`value()`](https://github.com/jhpyle/docassemble/blob/4ccd5afdab7efe21e156f7663791d1f6f8cc0214/docassemble_base/docassemble/base/functions.py#L4051) have a lot of overlap, and our `value_if_defined()` (which is equivalent to `showifdef()`) is essentially doing the same work twice.

There's less to do on our side, but I think we could:
* make a separate `ALAddendumFieldDict::has_overflow()` method that break out of the [`defined()` loop on the fields](https://github.com/SuffolkLITLab/docassemble-AssemblyLine/blob/4a9fa91d4a434392de92bd6e3330d16b691e9417/docassemble/AssemblyLine/al_document.py#L358).
* Dedup some of the work that's being done in `safe_value` and `overflow_value` when they're called from the same place (i.e. https://github.com/SuffolkLITLab/docassemble-AssemblyLine/blob/4a9fa91d4a434392de92bd6e3330d16b691e9417/docassemble/AssemblyLine/al_document.py#L107).

**Comment by nonprofittechy** [comment]
I'll take a look there first. I think most of the `value()` etc. functions are only triggered when an addendum is involved, and only for the fields that need to handle overflow, so I don't know if that could be the only explanation.

**Comment by nonprofittechy** [comment]
The speedup in #205 makes me wonder if it's worth using a pattern where every addendum field needs to be defined, or at least we have an explicit "optional" flag. That will be more confusing to use though.

**Comment by BryceStevenWilley** [comment]
Update: I've been trying the different ideas I had in the previous comment out, but none of them seem to be having any noticeable effect in the generation times. It could be a few things:
* The profiling I tried to do last time doesn't seem to actually catch much of the entire process: if DA has to stop and search for the definition of a variable at anytime, the profiling is just stopped and doesn't continue.  I do have a new way to profile which seems to catch much more of the whole processing step:
```
---
modules:
  - cProfile
  - pstats
---
objects:
  - download_file: DAFile
---
code: |
  profile = Profile()
---
code: |
  profile.enable()
  start_profile =True
---
code: |
  profile.disable()
  profile.dump_stats(download_file.path())
  del profile
  stop_profile = True
---
code: |
  download_file.initialize(filename='test.prof')
  new_stuff = True
---
# In the question that asks for the thing you're trying to time, for example, al_user_bundle, or tenant_repair_letter_bundle
id: download_conditions_checklist_docs
event: download_conditions_checklist_docs
question: |
  Your documents are ready  ${start_profile}
subquestion: |
  
  ${new_stuff}
  [prof](${ download_file.url_for() })
  ...
  ${ stop_profile }
---
```
* I just don't know which interview that the above results were on. I'm kicking myself right now for just not writing it down at all. It could be an outlier, or the few forms that I'm trying could be the outliers, I'm not sure.
* It could be the fact that I'm running things locally and not on a AWS hosted - S3 enabled Docassemble server.

Whichever it is, I ran the above code on Quinten's [Housing Code Checklist](https://github.com/nonprofittechy/docassemble-HousingCodeChecklist), which seems to be a good stress test: it takes ~18 seconds consistently to generate the final doc. I generated
[test.zip](https://github.com/SuffolkLITLab/docassemble-AssemblyLine/files/6756484/test.zip)
 on that interview, and it looks to be taking most of the time (75%) on `word_to_pdf()`. It still doesn't look like it's too much of an idempotency issue though, it's only called 8 times.
![Screenshot from 2021-07-02 14-58-53](https://user-images.githubusercontent.com/6252212/124318132-2e8b3200-db46-11eb-89b4-ce13a8418c2d.png)

I'm stuck trying to figure out why the `subprocess.run` in `word_to_pdf` is taking so long (~15 seconds, to convert 7 files to pdfs seems ridiculous).  But I don't think I can make much more progress here tbh. The other changes seem to work fine, but they just aren't giving the speed up I'd hoped for.

**Comment by nonprofittechy** [comment]
@BryceStevenWilley can we see how performance changes if we enable convertAPI or cloudconvert? I don't have an API key for either one but they both have free tiers.

https://docassemble.org/docs/config.html#libreoffice

**Comment by nonprofittechy** [comment]
I have suspected that starting up LibreOffice to convert files could be pretty slow. I think paying for an API (especially as expensive as the two options Docassemble supports) is probably non-starter, but maybe we could run LibreOffice in a listen mode instead of starting a new process for each conversion and assign a Lightsail instance or ec2 to run just conversions.

**Comment by BryceStevenWilley** [comment]
Welp, I tried it once with convertAPI, it took ~42 seconds to convert the documents, so I don't think either api will make it much faster tbh. 

I need to look into what libreoffice is doing more I think. It's trying to use [a macro](https://github.com/jhpyle/docassemble/blob/fd8b088c0c21f3c8a946d7e304932e9e08fa9a0d/docassemble_base/docassemble/base/pandoc.py#L372) to convert the files, but I can't actually get that macro to run outside of docassemble on the command line. Libreoffice in general shouldn't be too slow: there's another option to convert to PDF (I don't know if it's equivalent to what the macro is doing), `libreoffice --convert-to pdf my_docx_file.docx` takes no where near that time to convert several documents
```
brycew@fossa 15:38:22:~/Developer $ time libreoffice --convert-to pdf *.docx # converting 3 docxs

real	0m0.482s
user	0m0.005s
sys	0m0.005s
brycew@fossa 15:38:28:~/Developer $ time libreoffice --convert-to pdf Evidence_packet.docx

real	0m0.217s
user	0m0.006s
sys	0m0.008s
```
It looks like this was used before (there's still a commented out line with the same command), so I might ask Johnathan to see why that was changed.

**Comment by nonprofittechy** [comment]
I have a vague memory of some macro support being added to help update things like dynamic TOCs. Not sure if that's related--if so, maybe it should be an option in the attachment block.

I think this is the base of Jonathan's macro--at least the name seems to come from here: https://www.xml.com/pub/a/2006/01/11/from-microsoft-to-openoffice.html

**Comment by nonprofittechy** [comment]
Good job finding the actual macro: https://github.com/jhpyle/docassemble/blob/master/docassemble_base/docassemble/base/data/macros/Module1.xba

Noticing that it contains 4 `Wait`s totaling 1600 milliseconds. So maybe we can optimize that macro...

**Comment by nonprofittechy** [comment]
This idea is a little far out, but when I used to run Linux as my desktop I preferred AbiWord for every day use--it was much faster. It has a convert to PDF feature. An idea to keep in the back pocket.

I was also thinking re: your profiling, I greatly reduced the number of calls to `value()` so maybe that threw off your old profile?

**Comment by nonprofittechy** [comment]
Per Jonathan:

> One problem might be that I set it up with a spinlock, so only one instance of LibreOffice can run at one time in a container.  I did this because I was encountering difficult issues where LibreOffice, though launched in "headless" mode, was still acting like a desktop application, so that if you tried to run LibreOffice while LibreOffice was already running under the same user account, it wouldn't start.  It might be the case that the recent version of LibreOffice doesn't have that problem, so you could experiment with turning the spinlock off.
> I might have tried the listen mode thing and couldn't get it to work, although I could try it again now that LibreOffice is several versions ahead of where it was the last time I worked on the issue.

**Comment by nonprofittechy** [comment]
Notes from our 7/7/21 call: 

Why is the built-in file generation faster w/o ALDocument?
Is the difference as big w/ idempotency fixes, caching, and reduced calls to value()?
Problems w/ the `value()` function
Could we use the all_variables() function to basically get a cache of all of the lookups? Is that slower or faster?
Bryce is working on some fixes that should have at least a small improvement but they need more testing to do a PR.
Would using mako templates instead of DOCX be faster?
Do we need to do a new fork for libreoffice each time? Can it just be waiting?
Look at Jonathan’s macro
Can we remove waits? Is the retry 3 times enough to handle whatever the waits were added to accomplish?
Can we make the slow maco optional?

**Comment by BryceStevenWilley** [comment]
DA's [1.2.89 update](https://github.com/jhpyle/docassemble/compare/v1.2.88...v1.2.89#diff-3e8f4234e83187c13981d1c07a3e6aad9b9d8dc8507a240846113e0bc4395c72R56) removed those waits, which is exciting. It does bring down the times quite a bit! I tested the same interview generation before and after the update, and it went from ~1.77 seconds per call to `word_to_pdf()` to ~0.91 seconds per call, so about 2x speed up!

Like I mentioned before, it does seem like ALDocument is just a stress test of a lot of different components: each time I try to profile a different interview, I get a different function that's taking the most time. First it was `showifdef`, then it was `word_to_pdf`, and on my most recent profiling, it's `pdf_concatenate` that's taking the longest (which makes sense, because I added a lot of images and extra documents where I could this last time, and more of them were probably originally PDFs. There were 25x calls to `pdf_concatenate()`, vs only 11 last time). 

This "investigation" ticket may have reached it's end, but IMO we should focus on the background assembly, and then slowly chip away at what makes things take so long. I can make a ticket to properly triage all the different ways that assembling documents can be slow, and see what could improve things.
